backend/service-user/src/test/java/g55/cs3219/backend/userService/controller/UserControllerTest.java

```diff
@@ -0,0 +1,395 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class UserControllerTest {
+
+    @Mock
+    private UserService userService;
+
+    @Mock
+    private AuthenticationService authenticationService;
+
+    @Mock
+    private Authentication authentication;
+
+    @InjectMocks
+    private UserController userController;
+
+    private User adminUser;
+    private User regularUser;
+    private List<User> userList;
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
+
+        adminUser = new User("admin", "admin@example.com", "password");
+        adminUser.setId(1L);
+        adminUser.setAdmin(true);
+
+        regularUser = new User("user", "user@example.com", "password");
+        regularUser.setId(2L);
+        regularUser.setAdmin(false);
+
+        userList = Arrays.asList(adminUser, regularUser);
+    }
+
+    @Test
+    void getAllUsers_shouldReturnAllUsers_whenAdmin() {
+
+        when(authentication.getPrincipal()).thenReturn(adminUser);
+        when(userService.getAllUsers()).thenReturn(userList);
+
+        ResponseEntity<List<User>> response = userController.getAllUsers(authentication);
+
+        assertEquals(HttpStatus.OK, response.getStatusCode());
+        assertEquals(2, response.getBody().size());
+        verify(userService, times(1)).getAllUsers();
+    }
+
+    @Test
+    void getAllUsers_shouldReturnForbidden_whenNotAdmin() {
+        when(authentication.getPrincipal()).thenReturn(regularUser);
+
+        ResponseEntity<List<User>> response = userController.getAllUsers(authentication);
+
+        assertEquals(HttpStatus.FORBIDDEN, response.getStatusCode());
+        verify(userService, times(0)).getAllUsers();
+    }
+
+    @Test
+    void getUser_shouldReturnUser_whenAdmin() {
+        when(authentication.getPrincipal()).thenReturn(adminUser);
+        when(userService.getUserById(anyLong(), eq(adminUser))).thenReturn(regularUser);
+
+        ResponseEntity<?> response = userController.getUser("2", authentication);
+
+        assertEquals(HttpStatus.OK, response.getStatusCode());
+        UserResponse userResponse = (UserResponse) response.getBody();
+        assertNotNull(userResponse);
+        assertEquals(regularUser.getId(), userResponse.getId());
+        assertEquals(regularUser.getEmail(), userResponse.getEmail());
+        verify(userService, times(1)).getUserById(2L, adminUser);
+    }
+
+    @Test
+    void getUser_shouldReturnForbidden_whenRegularUserAccessingOtherUserData() {
+        when(authentication.getPrincipal()).thenReturn(regularUser);
+        doThrow(new RuntimeException("Forbidden")).when(userService).getUserById(anyLong(), eq(regularUser));
+
+        ResponseEntity<?> response = userController.getUser("1", authentication);
+
+        assertEquals(HttpStatus.FORBIDDEN, response.getStatusCode());
+        verify(userService, times(1)).getUserById(1L, regularUser);
+    }
+
+    @Test
```

Comment:
Great stuff! But we should probably change these exceptions thrown in the Services to checked exceptions, as we are expecting these errors to happen.

It'll also make more sense to throw the exceptions of a specific class, rather than matching the exceptions by the `getMessage()`